src/CanvasRenderingContext2d.cc

```diff
@@ -311,16 +363,27 @@ Context2d::setFillRule(v8::Local<v8::Value> value) {
 void
 Context2d::fill(bool preserve) {
   if (state->fillPattern) {
-    cairo_set_source(_context, state->fillPattern);
+    if (state->globalAlpha < 1) {
+      cairo_pattern_t *new_pattern = create_transparent_pattern(state->fillPattern, state->globalAlpha);
+      cairo_set_source(_context, new_pattern);
+    } else {
+      cairo_set_source(_context, state->fillPattern);
+    }
     cairo_pattern_set_extend(cairo_get_source(_context), CAIRO_EXTEND_REPEAT);
     // TODO repeat/repeat-x/repeat-y
   } else if (state->fillGradient) {
     cairo_pattern_set_filter(state->fillGradient, state->patternQuality);
-    cairo_set_source(_context, state->fillGradient);
+    if (state->globalAlpha < 1) {
+      cairo_pattern_t *new_gradient = create_transparent_gradient(state->fillGradient, state->globalAlpha);
+      cairo_pattern_set_filter(new_gradient, state->patternQuality);
+      cairo_set_source(_context, new_gradient);
+    } else {
+      cairo_pattern_set_filter(state->fillGradient, state->patternQuality);
+      cairo_set_source(_context, state->fillGradient);
+    }
   } else {
     setSourceRGBA(state->fill);
   }
-
```

Comment:
I can't comment far down enough, but after the `shadow`/`fill` calls a few lines down, you need to free the `new_pattern` or `new_gradient` `cairo_pattern_t`s with https://developer.gnome.org/cairo/stable/cairo-cairo-pattern-t.html#cairo-pattern-destroy.